src/librustc/ty/context.rs

```diff
@@ -917,6 +923,92 @@ pub struct GlobalCtxt<'tcx> {
     output_filenames: Arc<OutputFilenames>,
 }
 
+/// Everything needed to efficiently work with interned allocations
+#[derive(Debug, Default)]
+pub struct InterpretInterner<'tcx> {
+    /// Stores the value of constants (and deduplicates the actual memory)
+    allocs: FxHashSet<&'tcx interpret::Allocation>,
+
+    /// Allows obtaining function instance handles via a unique identifier
+    functions: FxHashMap<u64, Instance<'tcx>>,
+
+    /// Inverse map of `interpret_functions`.
+    /// Used so we don't allocate a new pointer every time we need one
+    function_cache: FxHashMap<Instance<'tcx>, u64>,
+
+    /// Allows obtaining const allocs via a unique identifier
+    alloc_by_id: FxHashMap<u64, &'tcx interpret::Allocation>,
+
+    /// The AllocId to assign to the next new regular allocation.
+    /// Always incremented, never gets smaller.
+    ///
+    /// A u64 will never overflow when being incremented.
```

Comment:
*confused* it will obviously overflow when incremented 2^64 times, so I must be just too blind to see the argument here...

Reply:
According to `2^64/3600/24/365.24/1024/1024/1024` you need to run `544` Giga-increments per second to overflow a `u64` in a year. It's essentially "never".

Reply:
Fair enough. :) But then please add that in the code... it wasn't clear to me that this is making such a time-based argument, and not, say, "address space is exhausted before this overflows".
(Also, doing a checked_add to increment so that we get panic on overflow wouldn't harm either... but that may just be my paranoid formal me that doesn't like arguments based on "not happening in practice" ;)